### ISSUE TITLE
always has some error

### ISSUE THREAD

**Issue by findagnar** [openzim/mwoffliner#407]
I think the project alway many error to stop you plan.→_→
And,Why check redirects one by one?

Caching Bottomless_Trap_Hole.jpg at /home/rot/zim/4/cac/moegirl_zh/m/a248b174e8d016c288da.jpg...
Executing command : jpegoptim --strip-all --force --all-normal -m60 "/home/rot/zim/4/cac/moegirl_zh/m/a248b174e8d016c288da.jpg"
Caching http://localhost:8000/zh.moegirl.org/v3/page/pagebundle/%E5%A5%A5%E7%89%B9%E6%9B%BC%3A%E9%AB%98%E6%96%AF?oldid=1874718 at /home/rot/zim/4/cac/moegirl_zh/3ad0c3e6efbaced68f0a...
Caching timg?image&quality=80&size=b9999_10000&sec=1508649045476&di=e7c0b78decb30abb72d24fe19063504c&imgtype=0&src=http%3A%2F%2Fa0.att.hudong.com%2F81%2F74%2F300001195481130336749987256_950.jpg at /home/rot/zim/4/cac/moegirl_zh/m/c52a8c34d1b934ceacc0...
Successfuly optimized /home/rot/zim/4/cac/moegirl_zh/m/a248b174e8d016c288da.jpg
Unable to create symlink to /home/rot/zim/4/tmp/moegirl_zh_all_2018-09/m/timg?image&quality=80&size=b9999_10000&sec=1508649045476&di=e7c0b78decb30abb72d24fe19063504c&imgtype=0&src=http://a0.att.hudong.com/81/74/300001195481130336749987256_950.jpg at /home/rot/zim/4/cac/moegirl_zh/m/c52a8c34d1b934ceacc0: Error: ENOENT: no such file or directory, symlink '/home/rot/zim/4/cac/moegirl_zh/m/c52a8c34d1b934ceacc0' -> '/home/rot/zim/4/tmp/moegirl_zh_all_2018-09/m/timg?image&quality=80&size=b9999_10000&sec=1508649045476&di=e7c0b78decb30abb72d24fe19063504c&imgtype=0&src=http://a0.att.hudong.com/81/74/300001195481130336749987256_950.jpg' true

**Comment by kelson42** [comment]
@findagnar There is a ticket about improving speed of redirect checks #325

**Comment by kelson42** [comment]
@findagnar What is the exact mwoffliner command you run?

**Comment by findagnar** [comment]
Do not use ---speed,the website will be tell me 429!!! **too much more**
So I think the api,Should be changed

> api.php?action=query&format=json&prop=revisions|coordinates|redirects&rawcontinue=1&generator=allpages&utf8=1&rvprop=ids|flags&colimit=max&gapfilterredir=nonredirects&gaplimit=max&gapnamespace=0&gapcontinue=

Also I think this bug appears in line 1936 of mwolliner.lib.js, 
> const mediaPath = getMediaPath(url);
you need to put the file name. Make a "/" to "%2f" replacement.

Sorry, machine translation language.

**Comment by kelson42** [comment]
@findagnar We still do not have your mwoffliner command line!

**Comment by findagnar** [comment]
I don't think the command line is heavy. After all, I have given the solution to the problem.


> mwoffliner

 --mwUrl=https://zh.moegirl.org --mwApiPath=api.php --adminEmail=*** --speed=0.7 --localParsoid --keepHtml --mwWikiPath=

![2ea4ffb553967cd2b0a959f8df8a5258](https://user-images.githubusercontent.com/16982413/46257854-cb32c800-c4f3-11e8-8ee3-250d1db7e121.jpg)

**Comment by findagnar** [comment]
@kelson42 
I'm sorry. I have to say this is a very unsuccessful program. I've been using it for two months and it makes me think it's very error-prone. Every mistake has to be made again. What broke me the most was that it was dead. Today is the second time.
Caching systems that allow users to take up a lot of disk space.
What I can't explain is that the redirect lookup has more than 100 TCP connections. And is downloading when 10.
In short, the program is always working on one CPU.
I would suggest you rewrite such a program... The logic of this program is too inefficient.



我很抱歉，我得说这是一个非常失败的程序。我用了两个月让我觉得，这个程序非常容易出错。每一次出错又得重新开始。最让我崩溃的是，它居然死掉无反应了。今天是第二次。
缓存系统，让使用者占用了非常多的磁盘空间。
令我无法解释的是，重定向查找有100多个tcp连接。而正在下载的时候10个一下。
总之程序总是在用一个cpu去处理。
挺建议你们重写一个这样的程序...这个程序的逻辑效率太低了。

**Comment by kelson42** [comment]
@findagnar I need to try and fix that problem indeed.

**Comment by findagnar** [comment]
@kelson42 
I don't think you're going to try to solve this whole bunch of problems.
Again, my idea is to rewrite the whole program in python. However, the idea is different. I am not sure that mwlib can achieve the effect of parsoid output web page. I don't have much time to test it.
Could you please leave your email? I want to communicate more.

我不觉得你会去试着解决这一堆问题。
再说，我的想法是用python重写整个程序。但思路不同，我不确定mwlib能不能做到parsoid输出网页的效果，我并没有太多时间去做测试。
能不能留下你的邮箱，我想多交流一些。

**Comment by findagnar** [comment]
@kelson42  One question the tool can build en_wikiepdia 6 million entries? 

mwoffliner --mwUrl=http://zh.moegirl.org --adminEmail=* --speed=0.4 --localParsoid --format=nopic --keepHtml --mwWikiPath= --mwApiPath=api.php

I think there are two problems with the first picture.
The third picture is nginx proxy configuration file.
![img20181020224959](https://user-images.githubusercontent.com/16982413/47257115-59c7b300-d4bc-11e8-9429-a7de54e0424f.jpg)
![img20181020230027](https://user-images.githubusercontent.com/16982413/47257121-6e0bb000-d4bc-11e8-9754-c78dba7c153a.jpg)
![img20181020230716](https://user-images.githubusercontent.com/16982413/47257161-f2f6c980-d4bc-11e8-9804-9432331e355d.jpg)

**Comment by findagnar** [comment]
@ISNIT0 
No...but the situation I encountered that time was actually more than these two problems.
I don't know how long it takes for 40,000 entries to rely on parsoid, through free cloudflare.
I am very sorry to have let the other server have crashed once.
1. Always use a cpu.
2. With the error, you will quit (anger)
3.tcp handshake failed error
4. Why can't I use sqlite as a data record?
5. Why don't you write a blog that clones en_wikipedia to warn users (I don't know how long it takes to clone an en_wikipedia)
This is the question I thought of two months ago. I don't know what the current situation is.

不...但我那次遇到的情况，实际上要比这两个问题还要多。
我不知道4万份词条依赖parsoid，通过免费的cloudflare，需要花费多长时间。
我已经感到非常抱歉，让对方服务器已经崩溃过一次了。
1.总是在使用一个cpu.
2.有了错误就会退出（愤怒）
3.tcp握手失败报错
4.为什么不可以使用sqlite作为数据记载
5.你们为什么不写个克隆en_wikipedia的博客来警告使用者（我不知道克隆一个en_wikipedia你们需要花上多长时间）
这是我两个月前所想到的问题。不知道现在的情况如何。

**Comment by ISNIT0** [comment]
Please correct me if I'm wrong - you'd like to report 4 concerns:

1. MWOffliner uses only one thread during scraping
2. MWOffliner quits after TCP Handshake failed errors
3. You'd like to be able to choose the datastore, rather than use Redis
4. You'd like more public information (in the form of blog posts) that talks about the process and technologies of scraping and reading MediaWiki sites (OpenZim/Kiwix)

My responses:

1. This is a good suggestion, there are various tools in the NodeJS ecosystem that would allow us to use multiple threads. This would be a very useful contribution from someone (I have opened a ticket: #429)
2. We have recently improved the error handling (on the master branch). If you see this issue, it would be helpful to get a stacktrace and which wiki you were scraping
3. If you have problems running Redis, you could use Docker. If you'd just like to use sqlite, you're welcome to contribute some code that facilitates this.
4. Blog posts take a long time to write, edit, and translate. If you'd like to volunteer your time I'm sure it'd be gratefully received (speak to @Popolechien)
Have you seen [download.kiwix.org](https://download.kiwix.org/zim/) which has various pre-made zim files?
